proposals/0529-th-quotes-as-patterns.rst

```diff
@@ -0,0 +1,356 @@
+The overall method of this proposal is to allow using them to solve more problems, so the AST becomes less necessary to use, and thus TH code in practice is less likely to break.
+
+Just a few lines above, however, there was more TH AST usage I couldn't get rid of::
+
+  let f :: forall d. Data d => d -> StateT [(Name, Exp)] Q d
+      f d = case eqT of
+        Just (Refl :: d :~: Exp)
+          | AppE (VarE m) eInner <- d
+          , m == 'unqMarker
+          -> do n <- lift $ newName "dynamicQuotedExpressionVariable"
+                modify ((n, eInner):)
+                return $ VarE n
+        _ -> gmapM f d
+  (e', exprsReversed) <- runStateT (gmapM f e) []
+
+Perhaps we can take solace in a claim that ``AppE`` and ``VarE`` are less likely to change, but that isn't satisfactory --- what about more complex patterns?
+The fundamental problem here is that in *positive* position (expressions), we have a choice of either using regular syntax or quotes,
+but in *negative* position (patterns), we only have the option regular syntax.
+
+The solution is simple: let's allow quotes too!
+This would allow:
+
+.. code-block:: diff
+
+         Just (Refl :: d :~: Exp)
+  -        | AppE (VarE m) eInner <- d
+  +        | [e| $(VarE m) $eInner |] <- d
+           , m == 'unqMarker
+           -> do ...
+
+or even going further:
+
+.. code-block:: diff
+
+         Just (Refl :: d :~: Exp)
+  -        | AppE (VarE m) eInner <- d
+  -        , m == 'unqMarker
+  +        | [e| unqMarker $eInner |] <- d
+           -> do ...
+
+In this way, we also avoid the use of the AST.
+
+Note we do have quotes *of patterns* today (``[p| ... |]``), but that is orthogonal.
+This is quotes *as patterns*, the type of syntax being quoted doesn't matter and could be anything.
+The point is the quotes are in negative position.
+
+With this change put together, the hope is that a significant portion of TH out in the wild is going to be more stable across GHC versions.
+This is all accomplished without trying to minimize TH AST changes, which is quite a hopeless task and also a perverse incentive for the rest of language development.
+
+Proposed Change Specification
+-----------------------------
+
+Quotes as patterns
+~~~~~~~~~~~~~~~~~~
+
+With the new extension ``TemplateHaskellQuotesAsPatterns``, slightly modified quotes are usable in pattern position.
+
+The first difference is that quotes as pattern match raw syntax, not (monadic) actions producing syntax.
+The tying rules are as follows:
+
+- ``[e| ... |]`` matches ``Exp``
+- ``[p| ... |]`` matches ``Pat``
+- ``[t| ... |]`` matches ``Type``
+- ``[d| ... |]`` matches ``Dec``
+
+The second differences is that splices within these quotes contain patterns instead of expressions::
+
+  p is in <pat>
+  --------------------------------
+  [| ... $(p) ... |] is in <apat>
+
+The third and final difference is that names in quotes must all be uses, never bindings.
```

Comment:
Presumably this means that matching on most top-level declarations is prohibited, as e.g. `[d| data T = MkT |]` binds names? Are there some examples where it is useful to match on a top-level declaration that does not bind any names?

I think you need to spell out the semantics of pattern-matching on name uses. I gather from the examples that you intend `[| name |]` to expand to a match on `VarE` plus an equality test with the in-scope `name`. But that should be written down explicitly as part of the Proposed Change Specification.

Reply:
I found 
```haskell
_ =...
```
as an example of a top level declaration that doesn't bind names
```haskell
instance $(t)
```
also

I will spell out.